pkg/controllermanager/controller/project/project_control_delete.go

```diff
@@ -74,6 +68,33 @@ func (c *defaultControl) deleteNamespace(project *gardencorev1beta1.Project, nam
 		return true, nil
 	}
 
-	err = c.k8sGardenClient.Client().Delete(context.TODO(), namespace, kubernetes.DefaultDeleteOptions...)
+	// If the user wants to keep the namespace in the system even if the project gets deleted then we remove the related
+	// labels, annotations, and owner references and only delete the project.
+	var keepNamespace bool
+	if val, ok := namespace.Annotations[common.NamespaceKeepAfterProjectDeletion]; ok {
+		keepNamespace, _ = strconv.ParseBool(val)
+	}
+
+	if keepNamespace {
+		delete(namespace.Annotations, common.NamespaceProject)
+		delete(namespace.Annotations, common.NamespaceProjectDeprecated)
+		delete(namespace.Annotations, common.NamespaceKeepAfterProjectDeletion)
+		delete(namespace.Labels, common.ProjectName)
+		delete(namespace.Labels, v1beta1constants.GardenRole)
+		delete(namespace.Labels, common.ProjectNameDeprecated)
+		delete(namespace.Labels, v1beta1constants.DeprecatedGardenRole)
+		for i := len(namespace.OwnerReferences) - 1; i >= 0; i-- {
+			if ownerRef := namespace.OwnerReferences[i]; ownerRef.APIVersion == gardencorev1beta1.SchemeGroupVersion.String() &&
+				ownerRef.Kind == "Project" &&
+				ownerRef.Name == project.Name &&
+				ownerRef.UID == project.UID {
+				namespace.OwnerReferences = append(namespace.OwnerReferences[:i], namespace.OwnerReferences[i+1:]...)
```

Comment:
Hmm, if you have multiple owner references that should be removed, this won't remove the second reference, but rather the one after that or result in an out of range panic, correct?
I know this will never happen, as the API server doesn't accept multiple owner references with the same field values. But for the sake of correct code, can you build up a slice of references that should be kept instead of removing items from the slice you're looping over? WDYT?

Reply:
AFAIK, this should work the way it is implemented because it just decreases by each iteration and doesn't use any index which is out of range.

Reply:
Oh well, you're right. I missed that 😄